<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Analyzis and filtering of phylogenomics datasets. 
	It takes an input either a collection of gene trees (then transformed to matrices) or directly a collection of gene matrices and performs an iterative process to identify
	what species in what genes are outliers, and whose elimination significantly improves the concordance between the input matrices. The methods builds upon the Distatis approach (Abdi et al. (2005) &lt;doi:10.1101/2021.09.08.459421&gt;), a generalization of classical multidimensional scaling to multiple distance matrices.">
<title>Detect and Remove Outliers in Phylogenomics Datasets • phylter</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Detect and Remove Outliers in Phylogenomics Datasets">
<meta property="og:description" content="Analyzis and filtering of phylogenomics datasets. 
	It takes an input either a collection of gene trees (then transformed to matrices) or directly a collection of gene matrices and performs an iterative process to identify
	what species in what genes are outliers, and whose elimination significantly improves the concordance between the input matrices. The methods builds upon the Distatis approach (Abdi et al. (2005) &lt;doi:10.1101/2021.09.08.459421&gt;), a generalization of classical multidimensional scaling to multiple distance matrices.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">phylter</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/runphylter.html">Use of phylter</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/damiendevienne/phylter/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="phylter-a-tool-for-analyzing-visualizing-and-filtering-phylogenomics-datasets-">PhylteR, a tool for analyzing, visualizing and filtering phylogenomics datasets <img src="reference/figures/logophylter.png" align="right" style="float:right; width:20%;"><a class="anchor" aria-label="anchor" href="#phylter-a-tool-for-analyzing-visualizing-and-filtering-phylogenomics-datasets-"></a>
</h1></div>
<p><a href="http://cran.r-project.org/package=phylter" class="external-link"><img src="http://www.r-pkg.org/badges/version/phylter" alt="CRAN_Status_Badge"></a> <a href="https://github.com/damiendevienne/phylter/actions" class="external-link"><img src="https://github.com/damiendevienne/phylter/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></p>
<p><strong><code>phylter</code></strong> is a tool that allows detecting, removing and visualizing outliers in phylogenomics dataset by iteratively removing taxa from gene families (gene trees) and optimizing a score of concordance between individual matrices.<br><strong><code>phylter</code></strong> relies on DISTATIS (Abdi et al, 2005), an extension of multidimensional scaling to 3 dimensions to compare multiple distance matrices at once.<br><strong><code>phylter</code></strong> builds on Phylo-MCOA (de Vienne et al. 2012) but is much faster and accurate.<br><strong><code>phylter</code></strong> takes as input either a collection of phylogenetic trees (that are converted to distance matrices by <code>phylter</code>), or a collection of pairwise distance matrices (obtained from multiple sequence alignements, for instance).<br><strong><code>phylter</code></strong> accepts data with missing values (missing taxa in some genes).<br><strong><code>phylter</code></strong> detects outliers with a method proposed by Hubert &amp; Vandervieren (2008) for skewed data.<br><strong><code>phylter</code></strong> does not accept that the same taxa is present multiple times in the same gene.</p>
<p><strong><code>phylter</code></strong> is written in R language.</p>
<p>For details about the functions, their usage, and a in-depth description of the use of <strong>phylter</strong> on a biological dataset, step-by-step, please vist the phylter web page : <a href="https://damiendevienne.github.io/phylter" class="external-link uri">https://damiendevienne.github.io/phylter</a>.</p>
<!-- > Note: if you don't use R or don't want to use R, a **containerized version of phylter** is also available: [https://wheredoweputthis.org](https://wheredoweputthis.org) -->
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong><code>phylter</code></strong> is not yet on CRAN (deposit in process). To install the development version:</p>
<ol style="list-style-type: decimal">
<li>Install the release version of <code>remotes</code> from CRAN:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Install the development version of <code>phylter</code> from GitHub:</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"damiendevienne/phylter"</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Once installed, the package can be loaded:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"phylter"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Note: phylter requires R version &gt; 4.0, otherwise it cannot be installed. Also, R uses the GNU Scientific Library. On Ubuntu, this can be installed prior to the installation of the phylter package by typing <code>sudo apt install libgsl-dev</code> in a terminal.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>Here is a brief introduction to the use <code>phylter</code> on a collection of gene trees. <!-- For more more detailed examples, please go to [ADD LINK TO THE AUTOMATICALLY GENERATED WEBSITEWEB](ADD LINK TO THE AUTOMATICALLY GENERATED WEBSITEWEB). --></p>
<p><strong>1.</strong> With the <code>read.tree</code> function from the <code>ape</code> package, read trees from external file and save as a list called <code>trees</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ape"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ape"</span><span class="op">)</span></span>
<span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="st">"treefile.tre"</span><span class="op">)</span></span></code></pre></div>
<p><strong>2.</strong> (optional) Read or get gene names somewhere (same order as the trees) and save it as a vector called <code>names</code>.</p>
<p><strong>3.</strong> Run <code>phylter</code> on your trees (see details below for possible options).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/phylter.html">phylter</a></span><span class="op">(</span><span class="va">trees</span>, gene.names <span class="op">=</span> <span class="va">names</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<h4 id="options">Options</h4>
<p>The <code>phylter</code> function is called as follows by default:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/phylter.html">phylter</a></span><span class="op">(</span><span class="va">X</span>, bvalue <span class="op">=</span> <span class="fl">0</span>, distance <span class="op">=</span> <span class="st">"patristic"</span>, k <span class="op">=</span> <span class="fl">3</span>, k2 <span class="op">=</span> <span class="va">k</span>, Norm <span class="op">=</span> <span class="st">"median"</span>, </span>
<span> Norm.cutoff <span class="op">=</span> <span class="fl">0.001</span>, gene.names <span class="op">=</span> <span class="cn">NULL</span>, test.island <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span> verbose <span class="op">=</span> <span class="cn">TRUE</span>, stop.criteria <span class="op">=</span> <span class="fl">1e-5</span>, InitialOnly <span class="op">=</span> <span class="cn">FALSE</span>, normalizeby <span class="op">=</span> <span class="st">"row"</span>, </span>
<span> parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Arguments are as follows:</p>
<ul>
<li>
<code>X</code>: A list of phylogenetic trees (phylo object) or a list of distance matrices. Trees can have different number of leaves and matrices can have different dimensions. If this is the case, missing values are imputed.</li>
<li>
<code>bvalue</code>: If <code>X</code> is a list of trees, nodes with a support below <code>bvalue</code> will be collapsed prior to the outlier detection.</li>
<li>
<code>distance</code>: If <code>X</code> is a list of trees, type of distance used to compute the pairwise matrices for each tree. Can be “patristic” (sum of branch lengths separating tips, the default) or “nodal” (number of nodes separating tips).</li>
<li>
<code>k</code>: Strength of outlier detection. The higher this value the less outliers detected.</li>
<li>
<code>k2</code>: Same as <code>k</code> for complete gene outlier detection. To preserve complete genes from being discarded, <code>k2</code> can be increased. By default, <code>k2 = k</code>.</li>
<li>
<code>Norm</code>: Should the matrices be normalized prior to the complete analysis and how. If “median”, matrices are divided by their median; if “mean”, they are divided by their mean; if “none”, no normalization if performed. Normalizing ensures that fast-evolving (and slow-evolving) genes are not treated as outliers. Normalization by median is a better choice as it is less sensitive to outlier values.</li>
<li>
<code>Norm.cutoff</code>: Value of the median (if <code>Norm = "median"</code>) or the mean (if <code>Norm = "mean"</code>) below which matrices are simply discarded from the analysis. This prevents dividing by 0, and allows getting rid of genes that contain mostly branches of length 0 and are therefore uninformative anyway. Discarded genes, if any, are listed in the output (<code>out$DiscardedGenes</code>).</li>
<li>
<code>gene.names</code>: List of gene names used to rename elements in <code>X</code>. If NULL (the default), elements are named 1,2,…,length(X).</li>
<li>
<code>test.island</code>: If <code>TRUE</code> (the default), only the highest value in an <em>island</em> of outliers is considered an outlier. This prevents non-outliers hitchhiked by outliers to be considered outliers themselves.</li>
<li>
<code>verbose</code>: If <code>TRUE</code> (the default), messages are written during the filtering process to get information on what is happening.</li>
<li>
<code>stop.criteria</code>: The optimization stops when the gain (quality of compromise) between round <em>n</em> and round <em>n</em>+1 is smaller than this value. Default to 1e-5.</li>
<li>
<code>InitialOnly</code>: Logical. If <code>TRUE</code>, only the Initial state of the data is computed.</li>
<li>
<code>normalizeby</code>: Should the gene x species matrix be normalized prior to outlier detection, and how.</li>
<li>
<code>parallel</code>: Logical. Should the computations be parallelized when possible? Default to <code>TRUE</code>. Note that the number of threads cannot be set by the user when <code>parallel = TRUE</code>. It uses all available cores on the machine.</li>
</ul>
</blockquote>
<p><strong>4.</strong> Analyze the results</p>
<p>To get the list of outliers detected by <code>phylter</code>, simply type:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span><span class="op">$</span><span class="va">Final</span><span class="op">$</span><span class="va">Outliers</span></span></code></pre></div>
<p>In addition, many functions allow looking at the outliers detected and comparing before and after <em>phy</em>ltering.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a summary: nb of outliers, gain in concordance, etc.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Show the number of species in each gene, and how many per gene are outliers</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span>, <span class="st">"genes"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Show the number of genes where each species is found, and how many are outliers</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span>, <span class="st">"species"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Compare before and after genes x species matrices, highlighting missing data and outliers </span></span>
<span><span class="co"># identified (not efficient for large datasets)</span></span>
<span><span class="fu"><a href="reference/plot.phylter.html">plot2WR</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot the dispersion of data before and after outlier removal. One dot represents one </span></span>
<span><span class="co"># gene x species association</span></span>
<span><span class="fu"><a href="reference/plot.phylter.html">plotDispersion</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot the genes x genes matrix showing pairwise correlation between genes</span></span>
<span><span class="fu"><a href="reference/plot.phylter.html">plotRV</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot optimization scores during optimization</span></span>
<span><span class="fu"><a href="reference/plot.phylter.html">plotopti</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span></code></pre></div>
<p><strong>5.</strong> Save the results of the analysis to an external file, for example to perform cleaning on raw alignments or pruning gene trees based on the results from <code>phylter</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/write.phylter.html">write.phylter</a></span><span class="op">(</span><span class="va">results</span>, file <span class="op">=</span> <span class="st">"phylter.out"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p>Abdi, H., O’Toole, A.J., Valentin, D. &amp; Edelman, B. (2005). <em>DISTATIS: The analysis of multiple distance matrices.</em> Proceedings of the IEEE Computer Society: International Conference on Computer Vision and Pattern Recognition (San Diego, CA, USA). doi: 10.1109/CVPR.2005.445. <a href="https://www.utdallas.edu/~herve/abdi-distatis2005.pdf" class="external-link uri">https://www.utdallas.edu/~herve/abdi-distatis2005.pdf</a></p></li>
<li><p>Allio, R., Tilak, M. K., Scornavacca, C., Avenant, N. L., Kitchener, A. C., Corre, E., … &amp; Delsuc, F. (2021). High-quality carnivoran genomes from roadkill samples enable comparative species delineation in aardwolf and bat-eared fox. Elife, 10, e63167. <a href="https://doi.org/10.7554/eLife.63167" class="external-link uri">https://doi.org/10.7554/eLife.63167</a></p></li>
<li><p>Hubert, M. and Vandervieren, E. (2008). <em>An adjusted boxplot for skewed distributions.</em> Computational Statistics and Data Analysis. <a href="https://doi.org/10.1016/j.csda.2007.11.008" class="external-link uri">https://doi.org/10.1016/j.csda.2007.11.008</a></p></li>
<li><p>de Vienne D.M., Ollier S. et Aguileta G. (2012). <em>Phylo-MCOA: A Fast and Efficient Method to Detect Outlier Genes and Species in Phylogenomics Using Multiple Co-inertia Analysis.</em> Molecular Biology and Evolution. <a href="https://doi.org/10.1093/molbev/msr317" class="external-link uri">https://doi.org/10.1093/molbev/msr317</a> (This is the ancestor of phylter).</p></li>
</ul>
<hr>
<p>For comments, suggestions and bug reports, please open an <a href="https://github.com/damiendevienne/phylter/issues" class="external-link">issue</a> on this GitHub repository.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/damiendevienne/phylter/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/damiendevienne/phylter/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 2)</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing phylter</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Damien M. de Vienne <br><small class="roles"> Author </small>  </li>
<li>Stéphane Dray <br><small class="roles"> Author </small>  </li>
<li>Aurélie Siberchicot <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Damien M. de Vienne, Stéphane Dray, Aurélie Siberchicot.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
