<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Functions to detect outliers, in matrices or in arrays."><title>Detection of outliers in 1D and 2D data — detect.outliers • phylter</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Detection of outliers in 1D and 2D data — detect.outliers"><meta property="og:description" content="Functions to detect outliers, in matrices or in arrays."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">phylter</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/runphylter.html">1-Use of phylter</a>
    <a class="dropdown-item" href="../articles/phyltercontainer.html">2-PhylteR in a container</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/damiendevienne/phylter/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Detection of outliers in 1D and 2D data</h1>
      <small class="dont-index">Source: <a href="https://github.com/damiendevienne/phylter/blob/HEAD/R/detect.outliers.R" class="external-link"><code>R/detect.outliers.R</code></a></small>
      <div class="d-none name"><code>detect.outliers.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Functions to detect outliers, in matrices or in arrays.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">detect.outliers</span><span class="op">(</span><span class="va">X</span>, k <span class="op">=</span> <span class="fl">3</span>, test.island <span class="op">=</span> <span class="cn">TRUE</span>, normalizeby <span class="op">=</span> <span class="st">"row"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">detect.outliers.array</span><span class="op">(</span><span class="va">arr</span>, <span class="va">nbsp</span>, k <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>X</dt>
<dd><p>2D matrix (gene x species) obtained with the <code><a href="Dist2WR.html">Dist2WR()</a></code> function.</p></dd>


<dt>k</dt>
<dd><p>strength of outlier detection. High values (typically 3) 
correspond to less outliers detected than with lower ones (e.g. 1.5).</p></dd>


<dt>test.island</dt>
<dd><p>should islands of outliers be treated as such. 
If TRUE (the default), only the highest value in an island of outliers is
removed. This prevents erroneous outliers due to hicthiking effect to be removed.</p></dd>


<dt>normalizeby</dt>
<dd><p>Should the 2D matrix be normalized prior to outlier detection, and how.
Can be "row" (the default),"col" or "none". Normalization is done by dividing columns or rows 
by their median.</p></dd>


<dt>arr</dt>
<dd><p>Array of values, typically the weight of each gene matrix (alpha values).</p></dd>


<dt>nbsp</dt>
<dd><p>Number of species in the analysis</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>detect.outliers</code>: A matrix with outliers detected in the 2D matric. Each row <code>x</code> contains the 
gene (<code>x[1]</code>) where the species (<code>x[2]</code>) is an outlier.</p>


<p><code>detect.outliers.array</code>: An array listing the outliers detected (if any)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>These functions detect outliers either in matrices or in arrays. For the method 
to be adapted to skewed data, as is the case here, the outlier detection 
method used is the adjusted Tukey proposed by Hubert and Vandervieren (2008).</p>
    </div>
    <div class="section level2">
    <h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a></h2>
    
<ul><li><p><code>detect.outliers()</code>: detect outliers in 2D matrix</p></li>
<li><p><code>detect.outliers.array()</code>: detects outliers in 1D array</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>de Vienne D.M., Ollier S. et Aguileta G. (2012) Phylo-MCOA: 
A Fast and Efficient Method to Detect Outlier Genes and Species
in Phylogenomics Using Multiple Co-inertia Analysis. Molecular 
Biology and Evolution 29 : 1587-1598.</p>
<p>Hubert, M. and Vandervieren, E. (2008). An adjusted boxplot for skewed
distributions. Computational Statistics and Data Analysis, 52, 5186-5201.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Get the initial gene x species matrix</span></span></span>
<span class="r-in"><span><span class="co"># from the carnivora dataset</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">carnivora</span><span class="op">)</span> </span></span>
<span class="r-in"><span><span class="va">mat</span><span class="op">&lt;-</span><span class="fu"><a href="phylter.html">phylter</a></span><span class="op">(</span><span class="va">carnivora</span>, InitialOnly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">WR</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Genes:    125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Species:  53</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># detect outliers in this matrix</span></span></span>
<span class="r-in"><span><span class="va">outliers</span><span class="op">&lt;-</span><span class="fu">detect.outliers</span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">outliers</span><span class="op">$</span><span class="va">cells</span> <span class="co"># matrix where each row represents one cell in the input matrix</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [,1] [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,]    2    6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,]    2   33</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,]    2   42</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,]   38    4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,]   38   53</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,]   38   30</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,]   52   48</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,]   55   28</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,]   56   31</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,]   60    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11,]   60   35</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12,]   60   38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13,]   60    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [14,]   64   33</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15,]   76    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16,]   76    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17,]   76   17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [18,]   76   33</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19,]   76   53</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [20,]   76    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [21,]   79    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [22,]   79   35</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [23,]   79    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [24,]   79    8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25,]   79   11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [26,]   81    5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [27,]   81   15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [28,]   81   38</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [29,]   81   11</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [30,]   84    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31,]   84    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [32,]   95    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [33,]   95   51</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [34,]  106   50</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [35,]  124   14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [36,]  124   16</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Damien M. de Vienne, Stéphane Dray, Théo Tricou, Aurélie Siberchicot.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

