<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Filter phylogenomics datasets — phylter • phylter</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Filter phylogenomics datasets — phylter"><meta name="description" content="Detection and filtering out of outliers in a list of trees
or a list of distance matrices."><meta property="og:description" content="Detection and filtering out of outliers in a list of trees
or a list of distance matrices."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phylter</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/runphylter.html">1-Use of phylter</a></li>
    <li><a class="dropdown-item" href="../articles/phyltercontainer.html">2-PhylteR in a container</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/damiendevienne/phylter/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Filter phylogenomics datasets</h1>
      <small class="dont-index">Source: <a href="https://github.com/damiendevienne/phylter/blob/master/R/phylter.R" class="external-link"><code>R/phylter.R</code></a></small>
      <div class="d-none name"><code>phylter.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Detection and filtering out of outliers in a list of trees
or a list of distance matrices.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">phylter</span><span class="op">(</span></span>
<span>  <span class="va">X</span>,</span>
<span>  bvalue <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  distance <span class="op">=</span> <span class="st">"patristic"</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  k2 <span class="op">=</span> <span class="va">k</span>,</span>
<span>  Norm <span class="op">=</span> <span class="st">"median"</span>,</span>
<span>  Norm.cutoff <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  gene.names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  test.island <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  stop.criteria <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  InitialOnly <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  normalizeby <span class="op">=</span> <span class="st">"row"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">X<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A list of phylogenetic trees (phylo object) or a list
of distance matrices. Trees can have different number of leaves and matrices
can have different dimensions. If this is the case, missing values are imputed.</p></dd>


<dt id="arg-bvalue">bvalue<a class="anchor" aria-label="anchor" href="#arg-bvalue"></a></dt>
<dd><p>If X is a list of trees, nodes with a support below 'bvalue' will be collapsed
prior to the outlier detection.</p></dd>


<dt id="arg-distance">distance<a class="anchor" aria-label="anchor" href="#arg-distance"></a></dt>
<dd><p>If X is a list of trees, type of distance used to compute the
pairwise matrices for each tree. Can be "patristic" (sum of branch lengths separating tips, the default)
or nodal (number of nodes separating tips). The "nodal" option should only be used if all species
are present in all genes.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>Strength of outlier detection. The higher this value the less outliers
detected (see details).</p></dd>


<dt id="arg-k-">k2<a class="anchor" aria-label="anchor" href="#arg-k-"></a></dt>
<dd><p>Same as k for complete gene outlier detection. To preserve complete genes from
being discarded, k2 can be increased. By default, k2 = k (see above).</p></dd>


<dt id="arg-norm">Norm<a class="anchor" aria-label="anchor" href="#arg-norm"></a></dt>
<dd><p>Should the matrices be normalized prior to the complete analysis and how. If "median" (the default), matrices are divided by their median, if
"mean" they are divided by their mean, if "none", no normalization if performed. Normalizing ensures that fast-evolving
(and slow-evolving) genes are not treated as outliers. Normalization by median is a better choice as it is less sensitive to outlier values.</p></dd>


<dt id="arg-norm-cutoff">Norm.cutoff<a class="anchor" aria-label="anchor" href="#arg-norm-cutoff"></a></dt>
<dd><p>Value of the median (if <code>Norm="median"</code>) or the mean (if
<code>Norm="mean"</code>) of phylogenetic distance matrices below which genes are simply discarded from the analysis. This
prevents dividing by 0, and allows getting rid of genes that contain mostly branches
of length 0 and are therefore uninformative anyway. Discarded genes, if any, are listed in
the output <code>out$DiscardedGenes</code>.</p></dd>


<dt id="arg-gene-names">gene.names<a class="anchor" aria-label="anchor" href="#arg-gene-names"></a></dt>
<dd><p>List of gene names used to rename elements in X. If NULL (the default), 0
elements are named 1,2,..,length(X).</p></dd>


<dt id="arg-test-island">test.island<a class="anchor" aria-label="anchor" href="#arg-test-island"></a></dt>
<dd><p>If TRUE (the default), only the highest value in
an 'island' of outliers is considered an outlier. This prevents non-outliers hitchhiked by outliers
to be considered outliers themselves.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>If TRUE (the default), messages are written during the filtering process to get information
of what is happening</p></dd>


<dt id="arg-stop-criteria">stop.criteria<a class="anchor" aria-label="anchor" href="#arg-stop-criteria"></a></dt>
<dd><p>The optimisation stops when the gain in concordance between matrices between round <code>n</code> and round <code>n+1</code> is smaller
than this value. Default to 1e-5.</p></dd>


<dt id="arg-initialonly">InitialOnly<a class="anchor" aria-label="anchor" href="#arg-initialonly"></a></dt>
<dd><p>Logical. If TRUE, only the Initial state of the data is computed.</p></dd>


<dt id="arg-normalizeby">normalizeby<a class="anchor" aria-label="anchor" href="#arg-normalizeby"></a></dt>
<dd><p>Should the gene x species matrix be normalized prior to outlier detection, and how.</p></dd>


<dt id="arg-parallel">parallel<a class="anchor" aria-label="anchor" href="#arg-parallel"></a></dt>
<dd><p>Should the computations be parallelized when possible? Default to TRUE. Note that the number of threads
cannot be set by the user when `parallel=TRUE`. It uses all available cores on the machine.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of class 'phylter' with the 'Initial' (before filtering) and 'Final' (after filtering) states,
or a list of class 'phylterinitial' only, if InitialOnly=TRUE. The function also returns the list of DiscardedGenes, if any.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">carnivora</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># using default paramaters</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">phylter</span><span class="op">(</span><span class="va">carnivora</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># perform the phylter analysis</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Genes:    125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Species:  53</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Initial score: 0.86235</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     28  new cells to remove -&gt; New score: 0.90272 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     18  new cells to remove -&gt; New score: 0.90833 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     16  new cells to remove -&gt; New score: 0.91501 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     18  new cells to remove -&gt; New score: 0.92561 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     5  new cells to remove -&gt; New score: 0.93404 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4  new cells to remove -&gt; New score: 0.93692 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     2  new cells to remove -&gt; New score: 0.93712 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1  new cells to remove -&gt; New score: 0.94392 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1  new cells to remove -&gt; New score: 0.94417 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     1  new cells to remove -&gt; New score: 0.94426 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  =&gt; No more outliers detected  -&gt;  Checking for complete gene outliers</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  =&gt; No more outliers detected  -&gt;  STOPPING OPTIMIZATION</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of outliers detected: 94</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of complete gene outliers : 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of complete species outliers : 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gain (concordance between matrices): 8.19% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Loss (data filtering): 1.42% </span>
<span class="r-in"><span><span class="va">res</span> <span class="co"># brief summary of the analysis</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylter Analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of class phylter</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call: phylter(X = carnivora, parallel = FALSE)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Initial	Initial matrices and values, before optimization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $Final		Final matrices, scores, outliers, after optimization</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $DiscardedGenes	List of discarded genes (not analyzed by phylter)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Tips:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Use summary(x) to get an overview of the results.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Use plot(x) to see the distribution of outliers</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Use plot2WR(x) to compare WR matrices before and after</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Use write.phylter(x) to write the results to an easily parsable file</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Use plotDispersion(x) to compare Distatis projections before and after</span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">DiscardedGenes</span> <span class="co"># list of genes discarded prior to the analysis</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> character(0)</span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">Initial</span> <span class="co"># See all elements prior to the analysis</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylter Analysis - initial state</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of class phylterinitial</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Object      Dimension Content                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 $mat.data   125       List of original distance matrices, one per gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 $WR         53 x 125  Species x Genes reference matrix                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 $RV         125 x 125 Genes x Genes RV correlation coefficients matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 $weights    125       Weight of each gene in the compromise           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 $compromise 53 x 53   Species x Species compromise matrix             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 $F          53 x 6    Distatis coordinates of compromise              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7 $matrices   125       Distatis coordinates of gene matrices (list)    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8 $PartialF   125       Species x Species gene matrices (list)          </span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">Final</span> <span class="co"># See all elements at the end of the analysis</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phylter Analysis - final state</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of class phylterfinal</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Object            Dimension</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  $WR               53 x 125 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  $RV               125 x 125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  $weights          125      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  $compromise       53 x 53  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  $F                53 x 8   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  $PartialF         125      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  $species.order    53       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  $AllOptiScores    11       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  $CELLSREMOVED     94 x 2   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 $Outliers         94 x 2   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 $CompleteOutliers 2        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 $matrices         125      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Content                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  Species x Genes reference matrix                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  Genes x Genes RV correlation coefficients matrix  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  Weight of each gene in the compromise             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  Species x Species compromise matrix               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  Distatis coordinates of compromise                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  Distatis coordinates of gene matrices (list)      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7  Name and order of species                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  Evolution of quality of compromise (11 steps)     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9  Index of cells removed (may contain imputed cells)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 Outliers detected (one row = one outlier cell)    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 Complete outliers (Gene and Species, if any)      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 Species x Species gene matrices (list)            </span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">Final</span><span class="op">$</span><span class="va">Outliers</span> <span class="co"># Print all outliers detected</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       [,1]                      [,2]                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] "ENSG00000005381_MPO"     "Arctocephalus_gazella"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,] "ENSG00000005381_MPO"     "Ursus_maritimus"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,] "ENSG00000005381_MPO"     "Ailurus_fulgens"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,] "ENSG00000106511_MEOX2"   "Panthera_tigris"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,] "ENSG00000106511_MEOX2"   "Mustela_putorius"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,] "ENSG00000114686_MRPL3"   "Procyon_lotor"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,] "ENSG00000116157_GPX7"    "Vulpes_vulpes"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,] "ENSG00000116761_CTH"     "Otocyon_megalotis"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,] "ENSG00000120053_GOT1"    "Phoca_vitulina"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,] "ENSG00000120053_GOT1"    "Gulo_gulo"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11,] "ENSG00000123307_NEUROD4" "Arctocephalus_gazella"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [12,] "ENSG00000132254_ARFIP2"  "Panthera_leo"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13,] "ENSG00000132254_ARFIP2"  "Odobenus_rosmarus"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [14,] "ENSG00000132254_ARFIP2"  "Felis_catus"               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15,] "ENSG00000132254_ARFIP2"  "Arctocephalus_gazella"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [16,] "ENSG00000132693_CRP"     "Hyaena_hyaena"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17,] "ENSG00000132693_CRP"     "Pteronura_brasiliensis"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [18,] "ENSG00000132693_CRP"     "Neovison_vison"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [19,] "ENSG00000132693_CRP"     "Eumetopias_jubatus"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [20,] "ENSG00000133135_RNF128"  "Cryptoprocta_ferox"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [21,] "ENSG00000133135_RNF128"  "Panthera_tigris"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [22,] "ENSG00000133135_RNF128"  "Phoca_vitulina"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [23,] "ENSG00000134240_HMGCS2"  "Gulo_gulo"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [24,] "ENSG00000134240_HMGCS2"  "Spilogale_gracilis"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [25,] "ENSG00000138675_FGF5"    "Taxidea_taxus"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [26,] "ENSG00000138675_FGF5"    "Lutra_lutra"               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [27,] "ENSG00000143125_PROK1"   "Potos_flavus"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [28,] "ENSG00000149573_MPZL2"   "Paradoxurus_hermaphroditus"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [29,] "ENSG00000073111_MCM2"    "Paradoxurus_hermaphroditus"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [30,] "ENSG00000106511_MEOX2"   "Manis_javanica"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [31,] "ENSG00000106511_MEOX2"   "Acinonyx_jubatus"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [32,] "ENSG00000106511_MEOX2"   "Panthera_pardus"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [33,] "ENSG00000106511_MEOX2"   "Leptonychotes_weddellii"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [34,] "ENSG00000114686_MRPL3"   "Arctocephalus_gazella"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [35,] "ENSG00000116157_GPX7"    "Canis_familiaris"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [36,] "ENSG00000116157_GPX7"    "Otocyon_megalotis"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [37,] "ENSG00000116761_CTH"     "Vulpes_vulpes"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [38,] "ENSG00000132693_CRP"     "Enhydra_lutris"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [39,] "ENSG00000132693_CRP"     "Crocuta_Crocuta"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [40,] "ENSG00000132693_CRP"     "Arctocephalus_gazella"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [41,] "ENSG00000132693_CRP"     "Zalophus_californianus"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [42,] "ENSG00000133135_RNF128"  "Ursus_americanus"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [43,] "ENSG00000138675_FGF5"    "Mustela_putorius"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [44,] "ENSG00000138675_FGF5"    "Mellivora_capensis"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [45,] "ENSG00000143125_PROK1"   "Paradoxurus_hermaphroditus"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [46,] "ENSG00000149573_MPZL2"   "Mungos_mungo"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [47,] "ENSG00000106511_MEOX2"   "Panthera_onca"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [48,] "ENSG00000106511_MEOX2"   "Felis_catus"               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [49,] "ENSG00000106511_MEOX2"   "Enhydra_lutris"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [50,] "ENSG00000106511_MEOX2"   "Ailuropoda_melanoleuca"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [51,] "ENSG00000106511_MEOX2"   "Odobenus_rosmarus"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [52,] "ENSG00000106511_MEOX2"   "Paradoxurus_hermaphroditus"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [53,] "ENSG00000106511_MEOX2"   "Neofelis_nebulosa"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [54,] "ENSG00000114686_MRPL3"   "Taxidea_taxus"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [55,] "ENSG00000116157_GPX7"    "Lycaon_pictus"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [56,] "ENSG00000132693_CRP"     "Lutra_lutra"               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [57,] "ENSG00000133135_RNF128"  "Procyon_lotor"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [58,] "ENSG00000138675_FGF5"    "Enhydra_lutris"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [59,] "ENSG00000143125_PROK1"   "Nasua_narica"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [60,] "ENSG00000144355_DLX1"    "Acinonyx_jubatus"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [61,] "ENSG00000149573_MPZL2"   "Suricata_suricatta"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [62,] "ENSG00000149573_MPZL2"   "Helogale_parvula"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [63,] "ENSG00000106511_MEOX2"   "Lynx_pardinus"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [64,] "ENSG00000106511_MEOX2"   "Gulo_gulo"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [65,] "ENSG00000106511_MEOX2"   "Ursus_maritimus"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [66,] "ENSG00000106511_MEOX2"   "Eumetopias_jubatus"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [67,] "ENSG00000106511_MEOX2"   "Neomonachus_schauinslandi" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [68,] "ENSG00000106511_MEOX2"   "Potos_flavus"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [69,] "ENSG00000106511_MEOX2"   "Panthera_leo"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [70,] "ENSG00000106511_MEOX2"   "Prionailurus_bengalensis"  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [71,] "ENSG00000106511_MEOX2"   "Canis_familiaris"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [72,] "ENSG00000106511_MEOX2"   "Neovison_vison"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [73,] "ENSG00000114686_MRPL3"   "Pteronura_brasiliensis"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [74,] "ENSG00000116761_CTH"     "Canis_familiaris"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [75,] "ENSG00000133135_RNF128"  "Mellivora_capensis"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [76,] "ENSG00000133135_RNF128"  "Acinonyx_jubatus"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [77,] "ENSG00000138675_FGF5"    "Gulo_gulo"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [78,] "ENSG00000138675_FGF5"    "Neovison_vison"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [79,] "ENSG00000143125_PROK1"   "Gulo_gulo"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [80,] "ENSG00000149573_MPZL2"   "Cryptoprocta_ferox"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [81,] "ENSG00000106511_MEOX2"   "Lynx_canadensis"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [82,] "ENSG00000106511_MEOX2"   "Zalophus_californianus"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [83,] "ENSG00000106511_MEOX2"   "Mirounga_angustirostris"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [84,] "ENSG00000133135_RNF128"  "Pteronura_brasiliensis"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [85,] "ENSG00000143125_PROK1"   "Otocyon_megalotis"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [86,] "ENSG00000106511_MEOX2"   "Puma_concolor"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [87,] "ENSG00000106511_MEOX2"   "Callorhinus_ursinus"       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [88,] "ENSG00000143125_PROK1"   "Vulpes_vulpes"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [89,] "ENSG00000143196_DPT"     "Ailurus_fulgens"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [90,] "ENSG00000138675_FGF5"    "Pteronura_brasiliensis"    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [91,] "ENSG00000143125_PROK1"   "Lycaon_pictus"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [92,] "ENSG00000143125_PROK1"   "Manis_javanica"            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [93,] "ENSG00000116761_CTH"     "Lycaon_pictus"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [94,] "ENSG00000143196_DPT"     "Taxidea_taxus"             </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co"># Change the call to phylter to use nodal distances, instead of patristic: </span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">phylter</span><span class="op">(</span><span class="va">carnivora</span>, distance <span class="op">=</span> <span class="st">"nodal"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Genes:    125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Species:  53</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Initial score: 0.90793</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     4  new cells to remove -&gt; New score: 0.90917 -&gt; OK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  =&gt; No more outliers detected  -&gt;  Checking for complete gene outliers</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  =&gt; No more outliers detected  -&gt;  STOPPING OPTIMIZATION</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of outliers detected: 4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of complete gene outliers : 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Number of complete species outliers : 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gain (concordance between matrices): 0.12% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Loss (data filtering): 0.06% </span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Damien M. de Vienne, Stéphane Dray, Théo Tricou, Aurélie Siberchicot.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

