<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>2-PhylteR in a container • phylter</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2-PhylteR in a container">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phylter</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.12</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/runphylter.html">1-Use of phylter</a></li>
    <li><a class="dropdown-item" href="../articles/phyltercontainer.html">2-PhylteR in a container</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/damiendevienne/phylter/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>2-PhylteR in a container</h1>
                        <h4 data-toc-skip class="author">Damien de
Vienne</h4>
            
            <h4 data-toc-skip class="date">2025-07-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/damiendevienne/phylter/blob/v0.9.12/vignettes/phyltercontainer.Rmd" class="external-link"><code>vignettes/phyltercontainer.Rmd</code></a></small>
      <div class="d-none name"><code>phyltercontainer.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="phylter-in-a-container">PhylteR in a container<a class="anchor" aria-label="anchor" href="#phylter-in-a-container"></a>
</h2>
<p>In addition to the R package <strong>phylter</strong> available on
CRAN (<a href="https://CRAN.R-project.org/package=phylter" class="external-link uri">https://CRAN.R-project.org/package=phylter</a>) and on
GitHub (<a href="https://github.com/damiendevienne/phylter" class="external-link uri">https://github.com/damiendevienne/phylter</a>),
<em>containerized</em> versions of <strong>phylter</strong> (docker and
singularity) are also proposed.</p>
<p>This may ease the use of <strong>phylter</strong> on some computing
infrastructures (clusters) or for users reluctant to the R language.</p>
<p>The containers host python3 scrips allowing to easily run
<strong>phylter</strong> with the same options than with the R package,
but also perform additional tasks such as removing (pruning) outliers
from input trees and/or filtering out outlier sequences from (aligned)
sequence files (fasta format).</p>
<p>Using <strong>phylter</strong> from the container simply consists in
running the <code>phylter.py</code> function, specifying various options
such as the directory containing the gene trees (with -t), the job name
(with -j), etc.</p>
<p>The containers also contain a toy dataset of 255 Carnivora genes
trees and alignments from Allio et al. (2021) that will allow you to
test both the correct installation of the container(s), and the use of
the <code>phylter.py</code> function and its options.</p>
<div class="section level3">
<h3 id="run-phylter-with-docker">Run PhylteR with Docker<a class="anchor" aria-label="anchor" href="#run-phylter-with-docker"></a>
</h3>
<p>PhylteR is available as a <a href="https://docs.docker.com/" class="external-link">Docker</a> container: <a href="https://hub.docker.com/r/treecoutheo/phylter_docker" class="external-link uri">https://hub.docker.com/r/treecoutheo/phylter_docker</a>.</p>
<p>Here are the steps needed to use the docker version of
<strong>phylter</strong>:</p>
<blockquote>
<p>Warning: you may need administrator rights to use docker!</p>
</blockquote>
<ol style="list-style-type: decimal">
<li>Pull the latest version of <strong>phylter</strong> container from
the Docker Hub repository:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">sudo</span> docker pull treecoutheo/phylter_docker:latest</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Run <strong>phylter</strong> on the example Carnivora dataset</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">sudo</span> docker run <span class="at">-v</span> <span class="va">$PWD</span>:<span class="va">$PWD</span> <span class="at">-w</span> <span class="va">$PWD</span> treecoutheo/phylter_docker phylter.py <span class="at">-j</span> Carnivora_docker <span class="at">-t</span> /usr/container-data/trees</span></code></pre></div>
<ul>
<li>
<code>$PWD</code> specifies that you work in your local
<strong>P</strong>resent <strong>W</strong>orking
<strong>D</strong>irectory. This should not be changed.</li>
<li>
<code>-j</code> specifies the job name for this
<strong>phylter</strong> run (and thus the name of the output
folder)</li>
<li>
<code>-t</code> specifies the folder containing the input gene trees
in Newick format (one tree per file)</li>
</ul>
<p>The command above creates the directory <code>Carnivora_docker</code>
that will contain:</p>
<ul>
<li>
<code>PhylteR_all_tree_named</code>: a single file containing all
the trees with the gene ID preceding the newick.</li>
<li>
<code>phylter.out</code>: the phylter output file containing the log
of the run and the list of identified outliers.</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li>Run <strong>phylter</strong> on the example Carnivora dataset AND
prune trees AND remove outliers from fasta files</li>
</ol>
<p>You may want to run <strong>phylter</strong> and to subsequently
remove the identified outliers from both the gene trees and the
sequences files. For this to be performed, the sequence files must
contain the same name as the corresponding tree, minus the extension if
any. For example, a sequence file named
<code>ENSG00000274286_ADRA2B_final_align_NT.aln</code> will be matched
automatically to a tree file named
<code>ENSG00000274286_ADRA2B.treefile</code>: <code>phylter.py</code>
will identify the gene ID as being
<strong>ENSG00000274286_ADRA2B</strong>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">sudo</span> docker run <span class="at">-v</span> <span class="va">$PWD</span>:<span class="va">$PWD</span> <span class="at">-w</span> <span class="va">$PWD</span> treecoutheo/phylter_docker phylter.py <span class="at">-j</span> Carnivora_docker <span class="at">-t</span> /usr/container-data/trees <span class="at">-p</span> TRUE <span class="at">-s</span> /usr/container-data/alignments <span class="at">-g</span> TRUE</span></code></pre></div>
<ul>
<li>
<code>-p TRUE</code> specifies that the input trees should be pruned
by removing the outliers identified by <strong>phylter</strong>. A new
directorry containing the pruned trees is created (see after).</li>
<li>
<code>-s</code> specifies the directory containing the sequences
(see above for specifications regarding naming conventions). If this
directory is specified, the input sequences will be filtered by removing
the outliers identified by <strong>phylter</strong>. A new directorry
containing the filtered sequences is created (see after).</li>
<li>
<code>-g TRUE</code> specifies that a full report (in pdf) should be
produced (in addition to the default <code>phylter.out</code>
file).</li>
</ul>
<p>The command above generates, in addition to the two files described
in the previous example:</p>
<ul>
<li><p><code>Carnivora_docker/trees_PhylteR/</code>: a directory
containing the trees pruned from their outliers. Note that the number of
trees inside the directory can be lower than the number of trees used as
input in case all species from a gene tree have been identified as
outliers.</p></li>
<li><p><code>Carnivora_docker/seqs_PhylteR/</code>: a directory
containing sequences with outlier sequences filtered out. The number of
sequences files inside the directory can be lower than the number of
sequences files used as input. Again, the number of sequence files
inside the directory can be lower than the number of trees sequence
files in the input directory in case all species from a gene tree have
been identified as outliers.</p></li>
<li><p><code>report.pdf</code>: a PDF report containing a summary of the
results.</p></li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>Prune trees and filter out sequences AFTER the
<strong>phylter</strong> run.</li>
</ol>
<p>Instead of performing the <strong>phylter</strong> analysis and the
filtering of outliers at the same time, you can do it in multiple steps.
here is how, on the example dataset:</p>
<ul>
<li>Run phylter:</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">sudo</span> docker run <span class="at">-v</span> <span class="va">$PWD</span>:<span class="va">$PWD</span> <span class="at">-w</span> <span class="va">$PWD</span> treecoutheo/phylter_docker phylter.py <span class="at">-j</span> Carnivora_docker <span class="at">-t</span> /usr/container-data/trees</span></code></pre></div>
<p>The output file <code>phylter.out</code> will be used for performing
the pruning and/or the sequence filtering (see below).</p>
<ul>
<li>Prune trees with <code>prune_tree_outliers.R</code>:</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">sudo</span> docker run <span class="at">-v</span> <span class="va">$PWD</span>:<span class="va">$PWD</span> <span class="at">-w</span> <span class="va">$PWD</span> treecoutheo/phylter_docker prune_tree_outliers.R container-data_phylter /usr/container-data/trees Carnivora_docker/phylter.out</span></code></pre></div>
<ul>
<li>Filter alignments with
<code>remove_sequence_outliers.py</code>:</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">sudo</span> docker run <span class="at">-v</span> <span class="va">$PWD</span>:<span class="va">$PWD</span> <span class="at">-w</span> <span class="va">$PWD</span> treecoutheo/phylter_docker remove_sequence_outliers.py <span class="at">-j</span> container-data_phylter <span class="at">-s</span> /usr/container-data/alignments <span class="at">-o</span> Carnivora_docker/phylter.out</span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>View all <strong>phylter</strong> options</li>
</ol>
<p><code>phylter.py</code> allows specifying all the options available
in the R package. To see this list of options, simply use the
<code>-h</code> option:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">sudo</span> docker run <span class="at">-v</span> <span class="va">$PWD</span>:<span class="va">$PWD</span> <span class="at">-w</span> <span class="va">$PWD</span> treecoutheo/phylter_docker phylter.py <span class="at">-h</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-phylter-with-singularity">Run PhylteR with Singularity<a class="anchor" aria-label="anchor" href="#run-phylter-with-singularity"></a>
</h3>
<p>PhylteR is also available as a <a href="https://docs.sylabs.io/guides/3.0/user-guide/index.html" class="external-link">singularity</a>
container : (<a href="https://cloud.sylabs.io/library/theo.treecou/tool/phylter_singularity" class="external-link uri">https://cloud.sylabs.io/library/theo.treecou/tool/phylter_singularity</a>).
Here are instructions to install (or build) and run it:</p>
<ol style="list-style-type: decimal">
<li>Pull the latest version of <strong>phylter</strong> container from
the Sylabs repository:</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">sudo</span> singularity pull PhylteR.sif library://theo.treecou/tool/phylter_singularity:latest</span></code></pre></div>
<p>Alternatively, you can build a singularity image from the Docker Hub
repository:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">sudo</span> singularity pull PhylteR.sif docker://treecoutheo/phylter_docker:latest</span></code></pre></div>
<p>2.a Run <strong>phylter</strong> on the carnivora example
dataset:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">singularity</span> exec <span class="at">-B</span> <span class="va">$PWD</span> PhylteR.sif phylter.py <span class="at">-j</span> Carnivora_singularity <span class="at">-t</span> /usr/container-data/trees</span></code></pre></div>
<blockquote>
<p>Note: For more options Please, refer to the description of the docker
container to see how to use all the options available with the
<code>phylter.py</code> function.</p>
</blockquote>
<p>2.b Alternatively, you can open a console in the singularity
container as follows and use R in that console:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">singularity</span> shell <span class="at">-B</span> <span class="va">$PWD</span> PhylteR.sif</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="ex">R</span> <span class="co"># this launch the version of R from inside the singularity</span></span></code></pre></div>
<p>Then:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/damiendevienne/phylter" class="external-link">phylter</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">list_trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="st">"/usr/container-data/trees/ENSG*.treefile"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">list_trees</span>, <span class="fu">ape</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phylter.html">phylter</a></span><span class="op">(</span><span class="va">trees</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Allio, R., Tilak, M.K., Scornavacca, C., Avenant, N.L., Kitchener,
A.C., Corre, E., Nabholz, B. &amp; Delsuc, F. (2021). <em>High-quality
carnivoran genomes from roadkill samples enable comparative species
delineation in aardwolf and bat-eared fox</em>. eLife, 10, e63167. doi:
10.7554/eLife.63167.</li>
</ul>
<hr>
<p>For comments, suggestions and bug reports, please open an <a href="https://github.com/damiendevienne/phylter/issues" class="external-link">issue</a> on
this GitHub repository.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Damien M. de Vienne, Stéphane Dray, Théo Tricou, Aurélie Siberchicot.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
