<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="phylter">
<title>Use of phylter • phylter</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Use of phylter">
<meta property="og:description" content="phylter">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">phylter</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/phyltercontainer.html">2-PhylteR in a container</a>
    <a class="dropdown-item" href="../articles/runphylter.html">Use of phylter</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/damiendevienne/phylter/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Use of phylter</h1>
                        <h4 data-toc-skip class="author">Damien de
Vienne</h4>
            
            <h4 data-toc-skip class="date">2023-02-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/damiendevienne/phylter/blob/HEAD/vignettes/runphylter.Rmd" class="external-link"><code>vignettes/runphylter.Rmd</code></a></small>
      <div class="d-none name"><code>runphylter.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong><code>phylter</code></strong> is not yet on CRAN (deposit in
process). To install the development version:</p>
<ol style="list-style-type: decimal">
<li>Install the release version of <code>remotes</code> from CRAN:</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Install the development version of <code>phylter</code> from
GitHub:</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"damiendevienne/phylter"</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Once installed, the package can be loaded:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"phylter"</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code>phylter</code> requires R version &gt; 4.0, otherwise
it cannot be installed. Also, R uses the GNU Scientific Library. On
Ubuntu, this can be installed prior to the installation of the
<code>phylter</code> package by typing
<code>sudo apt install libgsl-dev</code> in a terminal.</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>Here is a brief introduction to the use <code>phylter</code> on a
collection of gene trees. To better understand how <code>phylter</code>
really works, go to the section entitled <a href="#how-phylter-works">How phylter works</a>. To see its usage on a
small biological dataset and view the different outputs of the tool, go
to <a href="#example">Example</a> section. A full list of functions in
the <code>phylter</code> package is <a href="https://damiendevienne.github.io/phylter/reference/index.html" class="external-link">here</a>.</p>
<p><strong>1.</strong> With the <code>read.tree</code> function from the
<code>ape</code> package, read trees from external file and save as a
list called <code>trees</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ape"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ape"</span><span class="op">)</span></span>
<span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="st">"treefile.tre"</span><span class="op">)</span></span></code></pre></div>
<p><strong>2.</strong> (optional) Read or get gene names somewhere (same
order as the trees) and save it as a vector called
<code>names</code>.</p>
<p><strong>3.</strong> Run <code>phylter</code> on your trees (see
details below for possible options).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phylter.html">phylter</a></span><span class="op">(</span><span class="va">trees</span>, gene.names <span class="op">=</span> <span class="va">names</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/phylter.html">phylter()</a></code> function is called as follows by
default:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/phylter.html">phylter</a></span><span class="op">(</span><span class="va">X</span>, bvalue <span class="op">=</span> <span class="fl">0</span>, distance <span class="op">=</span> <span class="st">"patristic"</span>, k <span class="op">=</span> <span class="fl">3</span>, k2 <span class="op">=</span> <span class="va">k</span>, Norm <span class="op">=</span> <span class="st">"median"</span>, </span>
<span>  Norm.cutoff <span class="op">=</span> <span class="fl">0.001</span>, gene.names <span class="op">=</span> <span class="cn">NULL</span>, test.island <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>, stop.criteria <span class="op">=</span> <span class="fl">1e-5</span>, InitialOnly <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  normalizeby <span class="op">=</span> <span class="st">"row"</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Arguments are as follows:</p>
<ul>
<li>
<code>X</code>: A list of phylogenetic trees (phylo object) or a
list of distance matrices. Trees can have different number of leaves and
matrices can have different dimensions. If this is the case, missing
values are imputed.</li>
<li>
<code>bvalue</code>: If <code>X</code> is a list of trees, nodes
with a support below <code>bvalue</code> will be collapsed prior to the
outlier detection.</li>
<li>
<code>distance</code>: If <code>X</code> is a list of trees, type of
distance used to compute the pairwise matrices for each tree. Can be
“patristic” (sum of branch lengths separating tips, the default) or
“nodal” (number of nodes separating tips).</li>
<li>
<code>k</code>: Strength of outlier detection. The higher this value
the less outliers detected.</li>
<li>
<code>k2</code>: Same as <code>k</code> for complete gene outlier
detection. To preserve complete genes from being discarded,
<code>k2</code> can be increased. By default, <code>k2 = k</code>.</li>
<li>
<code>Norm</code>: Should the matrices be normalized prior to the
complete analysis and how. If “median”, matrices are divided by their
median; if “mean”, they are divided by their mean; if “none”, no
normalization if performed. Normalizing ensures that fast-evolving (and
slow-evolving) genes are not treated as outliers. Normalization by
median is a better choice as it is less sensitive to outlier
values.</li>
<li>
<code>Norm.cutoff</code>: Value of the median (if
<code>Norm = "median"</code>) or the mean (if
<code>Norm = "mean"</code>) below which matrices are simply discarded
from the analysis. This prevents dividing by 0, and allows getting rid
of genes that contain mostly branches of length 0 and are therefore
uninformative anyway. Discarded genes, if any, are listed in the output
(<code>out$DiscardedGenes</code>).</li>
<li>
<code>gene.names</code>: List of gene names used to rename elements
in <code>X</code>. If NULL (the default), elements are named
1,2,…,length(X).</li>
<li>
<code>test.island</code>: If <code>TRUE</code> (the default), only
the highest value in an <em>island</em> of outliers is considered an
outlier. This prevents non-outliers hitchhiked by outliers to be
considered outliers themselves.</li>
<li>
<code>verbose</code>: If <code>TRUE</code> (the default), messages
are written during the filtering process to get information on what is
happening.</li>
<li>
<code>stop.criteria</code>: The optimization stops when the gain
(quality of compromise) between round <em>n</em> and round <em>n</em>+1
is smaller than this value. Default to 1e-5.</li>
<li>
<code>InitialOnly</code>: Logical. If <code>TRUE</code>, only the
Initial state of the data is computed.</li>
<li>
<code>normalizeby</code>: Should the gene <span class="math inline">\(\times\)</span> species matrix be normalized prior
to outlier detection, and how.</li>
<li>
<code>parallel</code>: Logical. Should the computations be
parallelized when possible? Default to <code>TRUE</code>. Note that the
number of threads cannot be set by the user when
<code>parallel = TRUE</code>. It uses all available cores on the
machine.</li>
</ul>
<p><strong>4.</strong> Analyze the results.</p>
<p>You can get a list of outliers by simply typing:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span><span class="op">$</span><span class="va">Final</span><span class="op">$</span><span class="va">Outliers</span></span></code></pre></div>
<p>In addition, many functions allow looking at the outliers detected
and comparing before and after <em>phy</em>ltering. All these functions
are detailed in the <a href="#example">Example</a> section.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a summary: nb of outliers, gain in concordance, etc.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Show the number of species in each gene, and how many per gene are outliers</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span>, <span class="st">"genes"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Show the number of genes where each species is found, and how many are outliers</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">results</span>, <span class="st">"species"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Compare before and after genes $\times$ species matrices, highlighting missing data and outliers </span></span>
<span><span class="co"># identified (not efficient for large datasets)</span></span>
<span><span class="fu"><a href="../reference/plot.phylter.html">plot2WR</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot the dispersion of data before and after outlier removal. One dot represents one </span></span>
<span><span class="co"># gene $\times$ species association</span></span>
<span><span class="fu"><a href="../reference/plot.phylter.html">plotDispersion</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot the genes $\times$ genes matrix showing pairwise correlation between genes</span></span>
<span><span class="fu"><a href="../reference/plot.phylter.html">plotRV</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot optimization scores during optimization</span></span>
<span><span class="fu"><a href="../reference/plot.phylter.html">plotopti</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span> </span></code></pre></div>
<p><strong>5.</strong> Save the results of the analysis to an external
file, for example to perform cleaning on raw alignments based on the
results from <code>phylter</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write.phylter.html">write.phylter</a></span><span class="op">(</span><span class="va">results</span>, file <span class="op">=</span> <span class="st">"phylter.out"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="how-phylter-works">How <code>phylter</code> works<a class="anchor" aria-label="anchor" href="#how-phylter-works"></a>
</h2>
<p>The <code>phylter</code> method, in its entirety, is depicted in <a href="#fig1-phyler">Figure 1</a>. It starts with K distance matrices
obtained from K orthologous genes families by either computing pairwise
distances (sum of branch lengths) between species in each gene family
tree, or directly from each gene family multiple sequence alignment
(MSA). All the matrices are given the same dimensionality, using the
mean values to impute missing data if any, and are then normalised by
dividing each matrix by its median or its mean value (default is
median). The normalisation by median prevents genes from fast- (resp.
slow-) evolving orthologous gene families to be erroneously considered
outliers, and is a better choice than a normalisation by the mean as it
is less affected by outlier values.</p>
<br><div style="text-align:center">
<p><a name="fig1-phyler"></a> <span style="color:blue"><b>Figure 1.</b>
Principle of the phylter method for identifying outliers in phylogenomic
datasets. The method relies on DISTATIS (grey block), an extension of
multidimensional scaling to three dimensions.</span>
<img src="figures/method_new4.png" style="width:700px"></p>
</div>
<p><br></p>
<p>From the K matrices obtained, an incremental process starts,
consisting in three main steps (1) comparison of the matrices with the
DISTATIS method (Abdi et al. 2005; Abdi et al. 2012), (2) detection of
gene outliers, and (3) evaluation of the impact of removing these gene
outliers on the overall concordance between the matrices. Note that we
refer to gene outliers as single genes in single species that do not
follow the general trend, while outlier gene families refer to sets of
orthologous genes for a group of species (also referred to as gene
trees) that do not agree with the other gene families. These steps are
repeated until no more gene outlier is detected, or until the removal of
the identified gene outliers does not increase the concordance between
the matrices more than a certain amount specified by the user. Before
finishing the optimization, <code>phylter</code> performs a last action
consisting in checking whether some outlier gene families still exist
despite the removal of outlier genes already performed. These outlier
gene families correspond to gene families whose lack of correlation with
others is not due to a few outliers but are globally not following the
trend. If outlier gene families are discarded there, the optimization
restarts as it may have unblocked the detection of other gene
outliers.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<div class="section level3">
<h3 id="running-phylter">Running phylter<a class="anchor" aria-label="anchor" href="#running-phylter"></a>
</h3>
<p>A <code>carnivora</code> dataset (small subset from Allio et
al. 2021) comprised of 125 gene families for 53 species (53 <span class="math inline">\(\times\)</span> 125 = 6625 genes in total) is
included in the package. To load it and test <code>phylter</code> on
it:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">carnivora</span>, package <span class="op">=</span> <span class="st">"phylter"</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phylter.html">phylter</a></span><span class="op">(</span><span class="va">carnivora</span><span class="op">)</span> <span class="co"># for example</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="exploring-the-results">Exploring the results<a class="anchor" aria-label="anchor" href="#exploring-the-results"></a>
</h3>
<div class="section level4">
<h4 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h4>
<p>Typing <code>summary(results)</code> gives the following
information:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Total number of outliers detected: 94</span></span>
<span><span class="co">##   Number of complete gene outliers : 0</span></span>
<span><span class="co">##   Number of complete species outliers : 0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Gain (concordance between matrices): 8.19% </span></span>
<span><span class="co">## Loss (data filtering): 1.42%</span></span></code></pre>
<p>We see that with default parameters on the small
<code>carnivora</code> dataset, 94 <em>gene outliers</em> were
identified. No complete gene outliers (or <em>outlier gene
families</em>) were detected, meaning that there are no gene families
totally uncorrelated with the rest. There is also no complete species
outliers, i.e. species whose position is very variable in the different
gene trees (those are often called rogue taxa).</p>
</div>
<div class="section level4">
<h4 id="detailed-output">Detailed output<a class="anchor" aria-label="anchor" href="#detailed-output"></a>
</h4>
<p>The variable called <code>results</code> is a large object (a list)
of class <code>phylter</code>.</p>
<p>It is divided into two subgroups (lists) defining the
<em>Initial</em> (<code>results$Initial</code>) and the <em>Final</em>
(<code>results$Final</code>) states for all the objects manipulated by
<code>phylter</code> (see <a href="#fig1-phyler">Figure 1</a>).</p>
<p>You can view the content of these lists and the description of each
object it contains, like this:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span><span class="op">$</span><span class="va">Initial</span></span></code></pre></div>
<pre><code><span><span class="co">## Phylter Analysis - initial state</span></span>
<span><span class="co">## List of class phylterinitial</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Object      Dimension Content                                         </span></span>
<span><span class="co">## 1 $mat.data   125       List of original distance matrices, one per gene</span></span>
<span><span class="co">## 2 $WR         53 x 125  Species x Genes reference matrix                </span></span>
<span><span class="co">## 3 $RV         125 x 125 Genes x Genes RV correlation coefficients matrix</span></span>
<span><span class="co">## 4 $weights    125       Weight of each gene in the compromise           </span></span>
<span><span class="co">## 5 $compromise 53 x 53   Species x Species compromise matrix             </span></span>
<span><span class="co">## 6 $F          53 x 6    Distatis coordinates of compromise              </span></span>
<span><span class="co">## 7 $matrices   125       Distatis coordinates of gene matrices (list)    </span></span>
<span><span class="co">## 8 $PartialF   125       Species x Species gene matrices (list)</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span><span class="op">$</span><span class="va">Final</span></span></code></pre></div>
<pre><code><span><span class="co">## Phylter Analysis - final state</span></span>
<span><span class="co">## List of class phylterfinal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##    Object            Dimension</span></span>
<span><span class="co">## 1  $WR               53 x 125 </span></span>
<span><span class="co">## 2  $RV               125 x 125</span></span>
<span><span class="co">## 3  $weights          125      </span></span>
<span><span class="co">## 4  $compromise       53 x 53  </span></span>
<span><span class="co">## 5  $F                53 x 8   </span></span>
<span><span class="co">## 6  $PartialF         125      </span></span>
<span><span class="co">## 7  $species.order    53       </span></span>
<span><span class="co">## 8  $AllOptiScores    11       </span></span>
<span><span class="co">## 9  $CELLSREMOVED     94 x 2   </span></span>
<span><span class="co">## 10 $Outliers         94 x 2   </span></span>
<span><span class="co">## 11 $CompleteOutliers 2        </span></span>
<span><span class="co">## 12 $matrices         125      </span></span>
<span><span class="co">##    Content                                           </span></span>
<span><span class="co">## 1  Species x Genes reference matrix                  </span></span>
<span><span class="co">## 2  Genes x Genes RV correlation coefficients matrix  </span></span>
<span><span class="co">## 3  Weight of each gene in the compromise             </span></span>
<span><span class="co">## 4  Species x Species compromise matrix               </span></span>
<span><span class="co">## 5  Distatis coordinates of compromise                </span></span>
<span><span class="co">## 6  Distatis coordinates of gene matrices (list)      </span></span>
<span><span class="co">## 7  Name and order of species                         </span></span>
<span><span class="co">## 8  Evolution of quality of compromise (11 steps)     </span></span>
<span><span class="co">## 9  Index of cells removed (may contain imputed cells)</span></span>
<span><span class="co">## 10 Outliers detected (one row = one outlier cell)    </span></span>
<span><span class="co">## 11 Complete outliers (Gene and Species, if any)      </span></span>
<span><span class="co">## 12 Species x Species gene matrices (list)</span></span></code></pre>
<p>As an example, you can access to the initial distance matrices by
typing <code>results$Initial$mat.data</code> or to the weights
associated to each gene by typing <code>results$Initial$weights</code>
(see <a href="#fig1-phyler">Figure 1</a>).</p>
<p>Maybe more interesting is to <strong>get access to the list of
outliers identified by <code>phylter</code></strong>. As we can see
above, this is done by typing <code>results$Final$Outliers</code>. This
prints a matrix with two columns (extract below). Each row is an
outlier. The first column represents the gene family were the outlier
was found and the second column is the species identified as outlier in
this gene family.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span><span class="op">$</span><span class="va">Final</span><span class="op">$</span><span class="va">Outliers</span></span></code></pre></div>
<pre><code><span><span class="co">##       [,1]                      [,2]                        </span></span>
<span><span class="co">##  [1,] "ENSG00000005381_MPO"     "Arctocephalus_gazella"     </span></span>
<span><span class="co">##  [2,] "ENSG00000005381_MPO"     "Ursus_maritimus"           </span></span>
<span><span class="co">##  [3,] "ENSG00000005381_MPO"     "Ailurus_fulgens"           </span></span>
<span><span class="co">##  [4,] "ENSG00000106511_MEOX2"   "Panthera_tigris"           </span></span>
<span><span class="co">##  [5,] "ENSG00000106511_MEOX2"   "Mustela_putorius"          </span></span>
<span><span class="co">##  [6,] "ENSG00000114686_MRPL3"   "Procyon_lotor"             </span></span>
<span><span class="co">##  [7,] "ENSG00000116157_GPX7"    "Vulpes_vulpes"             </span></span>
<span><span class="co">##  [8,] "ENSG00000116761_CTH"     "Otocyon_megalotis"         </span></span>
<span><span class="co">##  [9,] "ENSG00000120053_GOT1"    "Phoca_vitulina"            </span></span>
<span><span class="co">## [10,] "ENSG00000120053_GOT1"    "Gulo_gulo"                 </span></span>
<span><span class="co">## [11,] "ENSG00000123307_NEUROD4" "Arctocephalus_gazella"     </span></span>
<span><span class="co">## [12,] "ENSG00000132254_ARFIP2"  "Panthera_leo"              </span></span>
<span><span class="co">## [13,] "ENSG00000132254_ARFIP2"  "Odobenus_rosmarus"         </span></span>
<span><span class="co">## [14,] "ENSG00000132254_ARFIP2"  "Felis_catus"               </span></span>
<span><span class="co">## [15,] "ENSG00000132254_ARFIP2"  "Arctocephalus_gazella"     </span></span>
<span><span class="co">## [16,] "ENSG00000132693_CRP"     "Hyaena_hyaena"             </span></span>
<span><span class="co">## [17,] "ENSG00000132693_CRP"     "Pteronura_brasiliensis"    </span></span>
<span><span class="co">## [18,] "ENSG00000132693_CRP"     "Neovison_vison"            </span></span>
<span><span class="co">## [19,] "ENSG00000132693_CRP"     "Eumetopias_jubatus"        </span></span>
<span><span class="co">## [20,] "ENSG00000133135_RNF128"  "Cryptoprocta_ferox"        </span></span>
<span><span class="co">## [21,] "ENSG00000133135_RNF128"  "Panthera_tigris"           </span></span>
<span><span class="co">## [22,] "ENSG00000133135_RNF128"  "Phoca_vitulina"            </span></span>
<span><span class="co">## [23,] "ENSG00000134240_HMGCS2"  "Gulo_gulo"                 </span></span>
<span><span class="co">## [24,] "ENSG00000134240_HMGCS2"  "Spilogale_gracilis"        </span></span>
<span><span class="co">## [25,] "ENSG00000138675_FGF5"    "Taxidea_taxus"             </span></span>
<span><span class="co">## [26,] "ENSG00000138675_FGF5"    "Lutra_lutra"               </span></span>
<span><span class="co">## [27,] "ENSG00000143125_PROK1"   "Potos_flavus"              </span></span>
<span><span class="co">## [28,] "ENSG00000149573_MPZL2"   "Paradoxurus_hermaphroditus"</span></span>
<span><span class="co">## [29,] "ENSG00000073111_MCM2"    "Paradoxurus_hermaphroditus"</span></span>
<span><span class="co">## [30,] "ENSG00000106511_MEOX2"   "Manis_javanica"            </span></span>
<span><span class="co">## [31,] "ENSG00000106511_MEOX2"   "Acinonyx_jubatus"          </span></span>
<span><span class="co">## [32,] "ENSG00000106511_MEOX2"   "Panthera_pardus"           </span></span>
<span><span class="co">## [33,] "ENSG00000106511_MEOX2"   "Leptonychotes_weddellii"   </span></span>
<span><span class="co">## [34,] "ENSG00000114686_MRPL3"   "Arctocephalus_gazella"     </span></span>
<span><span class="co">## [35,] "ENSG00000116157_GPX7"    "Canis_familiaris"          </span></span>
<span><span class="co">## [36,] "ENSG00000116157_GPX7"    "Otocyon_megalotis"         </span></span>
<span><span class="co">## [37,] "ENSG00000116761_CTH"     "Vulpes_vulpes"             </span></span>
<span><span class="co">## [38,] "ENSG00000132693_CRP"     "Enhydra_lutris"            </span></span>
<span><span class="co">## [39,] "ENSG00000132693_CRP"     "Crocuta_Crocuta"           </span></span>
<span><span class="co">## [40,] "ENSG00000132693_CRP"     "Arctocephalus_gazella"     </span></span>
<span><span class="co">## [41,] "ENSG00000132693_CRP"     "Zalophus_californianus"    </span></span>
<span><span class="co">## [42,] "ENSG00000133135_RNF128"  "Ursus_americanus"          </span></span>
<span><span class="co">## [43,] "ENSG00000138675_FGF5"    "Mustela_putorius"          </span></span>
<span><span class="co">## [44,] "ENSG00000138675_FGF5"    "Mellivora_capensis"        </span></span>
<span><span class="co">## [45,] "ENSG00000143125_PROK1"   "Paradoxurus_hermaphroditus"</span></span>
<span><span class="co">## [46,] "ENSG00000149573_MPZL2"   "Mungos_mungo"              </span></span>
<span><span class="co">## [47,] "ENSG00000106511_MEOX2"   "Panthera_onca"             </span></span>
<span><span class="co">## [48,] "ENSG00000106511_MEOX2"   "Felis_catus"               </span></span>
<span><span class="co">## [49,] "ENSG00000106511_MEOX2"   "Enhydra_lutris"            </span></span>
<span><span class="co">## [50,] "ENSG00000106511_MEOX2"   "Ailuropoda_melanoleuca"    </span></span>
<span><span class="co">## [51,] "ENSG00000106511_MEOX2"   "Odobenus_rosmarus"         </span></span>
<span><span class="co">## [52,] "ENSG00000106511_MEOX2"   "Paradoxurus_hermaphroditus"</span></span>
<span><span class="co">## [53,] "ENSG00000106511_MEOX2"   "Neofelis_nebulosa"         </span></span>
<span><span class="co">## [54,] "ENSG00000114686_MRPL3"   "Taxidea_taxus"             </span></span>
<span><span class="co">## [55,] "ENSG00000116157_GPX7"    "Lycaon_pictus"             </span></span>
<span><span class="co">## [56,] "ENSG00000132693_CRP"     "Lutra_lutra"               </span></span>
<span><span class="co">## [57,] "ENSG00000133135_RNF128"  "Procyon_lotor"             </span></span>
<span><span class="co">## [58,] "ENSG00000138675_FGF5"    "Enhydra_lutris"            </span></span>
<span><span class="co">## [59,] "ENSG00000143125_PROK1"   "Nasua_narica"              </span></span>
<span><span class="co">## [60,] "ENSG00000144355_DLX1"    "Acinonyx_jubatus"          </span></span>
<span><span class="co">## [61,] "ENSG00000149573_MPZL2"   "Suricata_suricatta"        </span></span>
<span><span class="co">## [62,] "ENSG00000149573_MPZL2"   "Helogale_parvula"          </span></span>
<span><span class="co">## [63,] "ENSG00000106511_MEOX2"   "Lynx_pardinus"             </span></span>
<span><span class="co">## [64,] "ENSG00000106511_MEOX2"   "Gulo_gulo"                 </span></span>
<span><span class="co">## [65,] "ENSG00000106511_MEOX2"   "Ursus_maritimus"           </span></span>
<span><span class="co">## [66,] "ENSG00000106511_MEOX2"   "Eumetopias_jubatus"        </span></span>
<span><span class="co">## [67,] "ENSG00000106511_MEOX2"   "Neomonachus_schauinslandi" </span></span>
<span><span class="co">## [68,] "ENSG00000106511_MEOX2"   "Potos_flavus"              </span></span>
<span><span class="co">## [69,] "ENSG00000106511_MEOX2"   "Panthera_leo"              </span></span>
<span><span class="co">## [70,] "ENSG00000106511_MEOX2"   "Prionailurus_bengalensis"  </span></span>
<span><span class="co">## [71,] "ENSG00000106511_MEOX2"   "Canis_familiaris"          </span></span>
<span><span class="co">## [72,] "ENSG00000106511_MEOX2"   "Neovison_vison"            </span></span>
<span><span class="co">## [73,] "ENSG00000114686_MRPL3"   "Pteronura_brasiliensis"    </span></span>
<span><span class="co">## [74,] "ENSG00000116761_CTH"     "Canis_familiaris"          </span></span>
<span><span class="co">## [75,] "ENSG00000133135_RNF128"  "Mellivora_capensis"        </span></span>
<span><span class="co">## [76,] "ENSG00000133135_RNF128"  "Acinonyx_jubatus"          </span></span>
<span><span class="co">## [77,] "ENSG00000138675_FGF5"    "Gulo_gulo"                 </span></span>
<span><span class="co">## [78,] "ENSG00000138675_FGF5"    "Neovison_vison"            </span></span>
<span><span class="co">## [79,] "ENSG00000143125_PROK1"   "Gulo_gulo"                 </span></span>
<span><span class="co">## [80,] "ENSG00000149573_MPZL2"   "Cryptoprocta_ferox"        </span></span>
<span><span class="co">## [81,] "ENSG00000106511_MEOX2"   "Lynx_canadensis"           </span></span>
<span><span class="co">## [82,] "ENSG00000106511_MEOX2"   "Zalophus_californianus"    </span></span>
<span><span class="co">## [83,] "ENSG00000106511_MEOX2"   "Mirounga_angustirostris"   </span></span>
<span><span class="co">## [84,] "ENSG00000133135_RNF128"  "Pteronura_brasiliensis"    </span></span>
<span><span class="co">## [85,] "ENSG00000143125_PROK1"   "Otocyon_megalotis"         </span></span>
<span><span class="co">## [86,] "ENSG00000106511_MEOX2"   "Puma_concolor"             </span></span>
<span><span class="co">## [87,] "ENSG00000106511_MEOX2"   "Callorhinus_ursinus"       </span></span>
<span><span class="co">## [88,] "ENSG00000143125_PROK1"   "Vulpes_vulpes"             </span></span>
<span><span class="co">## [89,] "ENSG00000143196_DPT"     "Ailurus_fulgens"           </span></span>
<span><span class="co">## [90,] "ENSG00000138675_FGF5"    "Pteronura_brasiliensis"    </span></span>
<span><span class="co">## [91,] "ENSG00000143125_PROK1"   "Lycaon_pictus"             </span></span>
<span><span class="co">## [92,] "ENSG00000143125_PROK1"   "Manis_javanica"            </span></span>
<span><span class="co">## [93,] "ENSG00000116761_CTH"     "Lycaon_pictus"             </span></span>
<span><span class="co">## [94,] "ENSG00000143196_DPT"     "Taxidea_taxus"</span></span></code></pre>
<p>All the informations present in these different objects returned by
<code>phylter</code> can be difficult to fully understand and to easily
use for other steps of bioinformatics pipelines. For this reason,
multiple functions are proposed to either visualize the results or write
the results in formatted output files that can be used elsewhere. These
are detailed in the next sections.</p>
</div>
<div class="section level4">
<h4 id="visualize-the-correlation-between-gene-families-the-rv-matrix-before-and-after-phylter">Visualize the correlation between gene families (the RV matrix)
before and after <code>phylter</code><a class="anchor" aria-label="anchor" href="#visualize-the-correlation-between-gene-families-the-rv-matrix-before-and-after-phylter"></a>
</h4>
<p>As shown in <a href="#fig1-phyler">Figure 1</a>, at each step of the
iterative process, <code>phylter</code> computes a matrix (called RV)
containing the correlation coefficients between individual gene
matrices. This RV matrix thus gives information on the overall
congruence between each gene family and the others. With
<code>phylter</code>, we can visualize this matrix before
(<code>plotRV(results, what = "Initial")</code>) and after
(<code>plotRV(results, what = "Final")</code>) the detection and removal
of the outliers. The two matrices are displayed below. It clearly
appears that the removal of outliers identified by <code>phylter</code>
increased the overall concordance between the gene matrices by removing
the few genes that were causing this loss of concordance.</p>
<p><img src="figures/RVcoeffs.png" style="width:100.0%"></p>
</div>
<div class="section level4">
<h4 id="visualize-the-2-way-reference-matrix-from-which-outliers-are-detected">Visualize the 2-way reference matrix from which outliers are
detected<a class="anchor" aria-label="anchor" href="#visualize-the-2-way-reference-matrix-from-which-outliers-are-detected"></a>
</h4>
<p>The 2WR matrix (see <a href="#fig1-phyler">Figure 1</a>) is the
species <span class="math inline">\(\times\)</span> genes matrix
computed at each loop of the <code>phylter</code> iterative process,
from which outliers are detected. A large value in one cell of this
matrix (light blue cells in the following figure) refers to one species
in one gene tree whose position is not in accordance with its position
in the other gene trees. The two 2WR matrices below represent the 2WR
matrix before (left) and after (right) using <code>phylter</code>. In
yellow, we see the outliers that <code>phylter</code> identified and
discarded during the process. We observe that the cells with high values
on the left matrix were those identified as outliers by
<code>phylter</code>, which is expected. The fact that some cells on the
left matrix do not seem so different from the others but are still
identified as outlirs by <code>phylter</code> is because the process is
iterative and these less-obvious outliers were maybe more obvious after
having discarded a first set of outliers. This figure is obtained by
typing <code>plot2WR(results)</code>.</p>
<p><img src="figures/2WR.png" style="width:100.0%"></p>
</div>
<div class="section level4">
<h4 id="visualize-the-distribution-of-outliers">Visualize the distribution of outliers<a class="anchor" aria-label="anchor" href="#visualize-the-distribution-of-outliers"></a>
</h4>
<p>Typing <code>plot(results, "genes")</code> allows exploring the
number of outliers identified by <code>phylter</code> for each gene.
Here is how it looks for the example dataset:</p>
<p><img src="figures/pergenes.png" style="width:100.0%"></p>
<p>We observe that a few gene families have a lot of outliers, while
others have none. One gene family has even almost half of its leaves
(species) discarded.</p>
<p>Typing <code>plot(results, "species")</code> allows exploring the
number of genes for which each species was identified as outlier.</p>
<p><img src="figures/perspecies.png" style="width:100.0%"></p>
<p>Here we see that no species is particularly problematic. Almost each
species is an outlier in at least one gene family.</p>
</div>
<div class="section level4">
<h4 id="visualize-the-effect-of-phylter-in-terms-of-data-dispersion">Visualize the effect of <code>phylter</code> in terms of data
dispersion<a class="anchor" aria-label="anchor" href="#visualize-the-effect-of-phylter-in-terms-of-data-dispersion"></a>
</h4>
<p>Removing outliers in a dataset can be seen as a way of decreasing the
<em>dispersion</em> of data if they were to be represented on a 2D map.
With <code>phylter</code>, this can be easily viewed. For each species,
we can compute the distance, in the compromise space (see <a href="#fig1-phyler">Figure 1</a>), between its position in a given gene
tree and its reference position. This can be done along the <em>x</em>
and the <em>y</em> axes for example, and be compared before and after
the use of <code>phylter</code>. This is what is represented hereafter.
We observe a sharp decrease of the dispersal of the points after having
employed <code>phylter</code> on the dataset.</p>
<p>This figure is obtained by simply typing
<code>plotDispersion(results)</code>.</p>
<p><img src="figures/phylterdispersion.png" style="width:100.0%"></p>
</div>
<div class="section level4">
<h4 id="visualize-the-evolution-of-the-quality-score-of-the-compromise-matrix">Visualize the evolution of the quality score of the compromise
matrix<a class="anchor" aria-label="anchor" href="#visualize-the-evolution-of-the-quality-score-of-the-compromise-matrix"></a>
</h4>
<p>At each step of the iterative process of <code>phylter</code> (<a href="#fig1-phyler">Figure 1</a>), a <em>quality</em> score is computed
describing how much individual gene matrices are correlated. This score
increases when outliers are removed, and the gain in quality achieved by
removing the outliers at each step is one of the criteria for stopping
the overall iterative process. The evolution of this score can be
visualized with <code>phylter</code> by typing
<code>plotopti(results)</code>. The output is given below.</p>
<p><img src="figures/phylteropti.png" style="width:100.0%"></p>
<p>We see that on this dataset, <code>phylter</code> performed 11 steps
before stopping.</p>
</div>
</div>
<div class="section level3">
<h3 id="writing-the-output">Writing the output<a class="anchor" aria-label="anchor" href="#writing-the-output"></a>
</h3>
<p><code>phylter</code> can write two types of output files:</p>
<ul>
<li>a <strong>pdf report</strong>, containing a summary of the results
along with all the figures presented above. To get the pdf report,
simply type <code>write.phylter(results, pdfreport = TRUE)</code>. This
will generate a file called <code>report.pdf</code>. An example of such
a file can be can be seen <a href="report.pdf">here</a>.</li>
<li>a <strong>text report</strong> with a summary of the
<code>phylter</code> run performed and a formatted list of outliers. To
get a text report, easy to parse and to use with other scripts in order
(for instance) to clean the multiple sequence alignement or to prune
outliers in gene trees, you just have to type
<code>write.phylter(results, file = "output.txt")</code>. This will
generate a file called <code>output.txt</code> that will ressemble <a href="output.txt">this one</a>.</li>
</ul>
<p>If no file name is given, the report is written to the console. Such
an example is given below.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write.phylter.html">write.phylter</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li><p>Abdi, H., O’Toole, A.J., Valentin, D. &amp; Edelman, B. (2005).
<em>DISTATIS: The Analysis of Multiple Distance Matrices</em>. 2005 IEEE
Computer Society Conference on Computer Vision and Pattern Recognition
(CVPR’05) - Workshops. doi: 10.1109/CVPR.2005.445. <a href="https://www.utdallas.edu/~herve/abdi-distatis2005.pdf" class="external-link uri">https://www.utdallas.edu/~herve/abdi-distatis2005.pdf</a></p></li>
<li><p>Allio, R., Tilak, M.K., Scornavacca, C., Avenant, N.L.,
Kitchener, A.C., Corre, E., Nabholz, B. &amp; Delsuc, F. (2021).
<em>High-quality carnivoran genomes from roadkill samples enable
comparative species delineation in aardwolf and bat-eared fox</em>.
eLife, 10, e63167. doi: 10.7554/eLife.63167.</p></li>
<li><p>Hubert, M. and Vandervieren, E. (2008). <em>An adjusted boxplot
for skewed distributions</em>. Computational Statistics and Data
Analysis. doi: 10.1016/j.csda.2007.11.008.</p></li>
<li><p>de Vienne D.M., Ollier S. et Aguileta G. (2012). <em>Phylo-MCOA:
A Fast and Efficient Method to Detect Outlier Genes and Species in
Phylogenomics Using Multiple Co-inertia Analysis</em>. Molecular Biology
and Evolution. doi: 10.1093/molbev/msr317 (This is the ancestor of
phylter).</p></li>
</ul>
<hr>
<p>For comments, suggestions and bug reports, please open an <a href="https://github.com/damiendevienne/phylter/issues" class="external-link">issue</a> on
this GitHub repository.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Damien M. de Vienne, Stéphane Dray, Aurélie Siberchicot.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
